### PR TITLE
fix: render template load reset

### PR DESCRIPTION
**Pull request by butuzov**
This fixes reset for render's template load, bug appears only while
running tests, when previous one is changing text one (if none of
the template* params provided).

Adds tests for output.Option checkers

Also fixes two tests:
1) Undefined Interface - solution fixed golden
2) Entire Directory - solution provide template name (testify).

### CODE REVIEW

**Comment by butuzov** [comment]
It's weird that example 037 is passing, but honestly I have no idea without `-v` (which is gonna be added in #150 )

File: gotests_test.go [hunk 1]
```diff
@@ -492,6 +492,7 @@ func TestGenerateTests(t *testing.T) {
 			name: "Entire testdata directory",
 			args: args{
 				srcPath: `testdata/`,
+				template: "testify",
```
Comment (cweill) [inline gotests_test.go:495]: What happens when you remove this? Ideally, I'd like this these existing tests to not need to change. Especially this one which was added 5 years ago, before `template` was an argument.

File: internal/output/output.go [hunk 2]
```diff
@@ -24,19 +24,24 @@ type Options struct {
 	TemplateData   [][]byte
 }
 
+func (o *Options) isProvidesTemplateData() bool { return o != nil && len(o.TemplateData) > 0 }
```
Comment (cweill) [inline internal/output/output.go:27]: (nit) s/isProvides/provides/g here and throughout. For example ` func (o *Options) isProvidesTemplateData() bool` -> `func (o *Options) providesTemplateData() bool`.

**Review by cweill — CHANGES_REQUESTED** [review]

**Comment by cweill** [comment]
@butuzov You'll want to update the branch now.

File: gotests_test.go (hunk 1 above)
Comment (butuzov) [reply gotests_test.go:495]: I run it few times and this test generates (without pinned template)

```
 go test -v ./... -count 10 | grep TestGenerateTests/Entire_testdata_directory -A20 -B20


=== RUN   TestGenerateTests/Entire_testdata_directory
    gotests_test.go:846: "Entire testdata directory". GenerateTests(testdata/) error = output.Process: render.Header: template: no template "header" associated with template "render", wantErr false
--- FAIL: TestGenerateTests/Entire_testdata_directory (0.76s)
---------------------------------------------------------------
=== RUN   TestGenerateTests/Entire_testdata_directory
    gotests_test.go:846: "Entire testdata directory". GenerateTests(testdata/) error = output.Process: render.TestFunction: template: no template "function" associated with template "render", wantErr false
--- FAIL: TestGenerateTests/Entire_testdata_directory (0.64s)
---------------------------------------------------------------
=== RUN   TestGenerateTests/Entire_testdata_directory
--- PASS: TestGenerateTests/Entire_testdata_directory (2.49s)
---------------------------------------------------------------
=== RUN   TestGenerateTests/Entire_testdata_directory
    gotests_test.go:846: "Entire testdata directory". GenerateTests(testdata/) error = output.Process: render.TestFunction: template: no template "function" associated with template "render", wantErr false
--- FAIL: TestGenerateTests/Entire_testdata_directory (0.32s)
---------------------------------------------------------------
=== RUN   TestGenerateTests/Entire_testdata_directory
    gotests_test.go:846: "Entire testdata directory". GenerateTests(testdata/) error = output.Process: render.TestFunction: template: no template "function" associated with template "render", wantErr false
--- FAIL: TestGenerateTests/Entire_testdata_directory (0.36s)
---------------------------------------------------------------
=== RUN   TestGenerateTests/Entire_testdata_directory
    gotests_test.go:846: "Entire testdata directory". GenerateTests(testdata/) error = output.Process: render.Header: template: no template "header" associated with template "render", wantErr false
--- FAIL: TestGenerateTests/Entire_testdata_directory (0.37s)
---------------------------------------------------------------
=== RUN   TestGenerateTests/Entire_testdata_directory
    gotests_test.go:846: "Entire testdata directory". GenerateTests(testdata/) error = output.Process: render.Header: template: no template "header" associated with template "render", wantErr false
--- FAIL: TestGenerateTests/Entire_testdata_directory (0.72s)
---------------------------------------------------------------
=== RUN   TestGenerateTests/Entire_testdata_directory
    gotests_test.go:846: "Entire testdata directory". GenerateTests(testdata/) error = output.Process: render.Header: template: no template "header" associated with template "render", wantErr false
--- FAIL: TestGenerateTests/Entire_testdata_directory (0.98s)
---------------------------------------------------------------
=== RUN   TestGenerateTests/Entire_testdata_directory
    gotests_test.go:846: "Entire testdata directory". GenerateTests(testdata/) error = output.Process: render.Header: template: no template "header" associated with template "render", wantErr false
--- FAIL: TestGenerateTests/Entire_testdata_directory (0.69s)
---------------------------------------------------------------
=== RUN   TestGenerateTests/Entire_testdata_directory
    gotests_test.go:846: "Entire testdata directory". GenerateTests(testdata/) error = output.Process: render.TestFunction: template: render:65:15: executing "function" at <{{template "results" $f}}>: template "results" not defined, wantErr false
--- FAIL: TestGenerateTests/Entire_testdata_directory (1.08s)
```

File: internal/output/output.go (hunk 2 above)
Comment (butuzov) [reply internal/output/output.go:27]: I suck at naming things.

File: gotests_test.go (hunk 1 above)
Comment (butuzov) [reply gotests_test.go:495]: Hm... Did a stress test (`-count=10` ), it still fails sometimes at the same test

```
=== RUN   TestGenerateTests/Entire_testdata_directory
fatal error: concurrent map read and map write
```

So we have race under the stress of generateTests, should we address this now or add to the backlog?

**Comment by butuzov** [comment]
@cweill Did you have a chance to check my response? The test was added before the bug was introduced, even with this small fix, code has a race condition, I will fix if/when find.

File: gotests_test.go (hunk 1 above)
Comment (cweill) [reply gotests_test.go:565]: @butuzov Sorry for the delay. Ok I see. Let's add this race condition to the backlog.

**Review by cweill — APPROVED** [review]

**Comment by cweill** [comment]
@butuzov Thanks, sorry I missed it. Can you please file an issue to track this bug. Then you can `Squash and Merge`.

**Comment by butuzov** [comment]
> @butuzov Thanks, sorry I missed it. Can you please file an issue to track this bug. Then you can `Squash and Merge`.

I believe only collaborators can squash and merge. All I see is `Close with Comment` and `Comment`.  MAybe it also worth checking if `squash`'s enabled in settings.

PS
Issue Filed.

**Comment by butuzov** [comment]
Thanks. I will check what I can do with parallelization as it really affects tests atm. 
Cheers.

**Pull request merged**